llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -18200,6 +18200,19 @@ static SDValue performConcatVectorsCombine(SDNode *N,
     return true;
   };
 
+  // concat(uhadd(a, b), uhadd(c, d)) -> uhadd(concat(a, c), concat(b, d))
+  if (N0Opc == ISD::AVGFLOORU && N1Opc == ISD::AVGFLOORU) {
```

Comment:
I think we also need to check that the concat operation only has 2 operands. In theory, you can do concats of as many vectors as you want so long as they are all the same size.

Reply:
Do we also want to only perform the optimisation if there is just one use of both operands? i.e. you might have something like

  %u1 = uhadd a, b
  %u2 = uhadd b, c
  %c = concat(%u1, %u2)
  %u3 = add %u1, %u2

Reply:
Added the hasOneUse() check. As for checking the concat operation has 2 operands, since I moved the code it is now checked above.